# Agrupando e manipulando dados agrupados com uso do Pandas

É muito comum em Data Science nós termos de responder perguntas como:
- 'Qual a média salarial **por sexo**?'
- 'Qual a média salarial **por região do Brasil**?'
- 'Qual a nota máxima do ENEM **por ano**?'


Entre muitas outras similares.

Vemos que todas essas perguntas possuem uma similaridade, que é a parte do final: 'por sexo', 'por cor', 'por região', 'por idade', 'por cargo', 'por ano'... 
Em geral, queremos avaliar uma métrica de acordo de acordo com cada **categoria** ou **grupo**.

Para responder esse tipo de pergunta, precisamos **agrupar** nossos dados de acordo com cada categoria.

O *pandas* possui uma funcionalidade muito bacana para nos ajudar, que é a função *groupby*.

Primeiro, importamos o pandas, nosso foco hoje, e o numpy, que possivelmente vai nos ajudar mais para frente.

In [1]:
import numpy as np
import pandas as pd

Antes de mais nada, as versões do numpy e pandas que utilizei para fazer esse notebook são, respectivamente, 1.19.2 e 1.2.3.

Isso é importante, pois bibliotecas costumam mudar conforme o tempo e algo que eu mostrar aqui pode ficar ultrapassado no futuro.

In [2]:
print(f'A versão do numpy é : {np.__version__}')
print(f'A versão do pandas é : {pd.__version__}')

A versão do numpy é : 1.19.2
A versão do pandas é : 1.2.3


Vamos agora importar nossa tabela.

Vou utilizar uma tabela do censo americado de 2010 a 2015.

In [3]:
df = pd.read_csv('census.csv')
df

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME            CTYNAME  \
0         40       3         6      1       0  Alabama            Alabama   
1         50       3         6      1       1  Alabama     Autauga County   
2         50       3         6      1       3  Alabama     Baldwin County   
3         50       3         6      1       5  Alabama     Barbour County   
4         50       3         6      1       7  Alabama        Bibb County   
...      ...     ...       ...    ...     ...      ...                ...   
3188      50       4         8     56      37  Wyoming  Sweetwater County   
3189      50       4         8     56      39  Wyoming       Teton County   
3190      50       4         8     56      41  Wyoming       Uinta County   
3191      50       4         8     56      43  Wyoming    Washakie County   
3192      50       4         8     56      45  Wyoming      Weston County   

      CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  \
0           4779736            4780127          4785161  ...   
1             54571              54571            54660  ...   
2            182265             182265           183193  ...   
3             27457              27457            27341  ...   
4             22915              22919            22861  ...   
...             ...                ...              ...  ...   
3188          43806              43806            43593  ...   
3189          21294              21294            21297  ...   
3190          21118              21118            21102  ...   
3191           8533               8533             8545  ...   
3192           7208               7208             7181  ...   

      RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0             0.002295         -0.193196          0.381066          0.582002   
1             7.242091         -2.915927         -3.012349          2.265971   
2            14.832960         17.647293         21.845705         19.243287   
3            -4.728132         -2.500690         -7.056824         -3.904217   
4            -5.527043         -5.068871         -6.201001         -0.177537   
...                ...               ...               ...               ...   
3188          1.072643         16.243199         -5.339774        -14.252889   
3189         -1.589565          0.972695         19.525929         14.143021   
3190        -17.755986         -4.916350         -6.902954        -14.215862   
3191        -11.637475         -0.827815         -2.013502        -17.781491   
3192        -11.752361         -8.040059         12.372583          1.533635   

      RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0            -0.467369     1.030015     0.826644     1.383282     1.724718   
1            -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2            17.197872    15.844176    18.559627    22.727626    20.317142   
3           -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4             0.177258    -5.088389    -4.363636    -5.403729     0.754533   
...                ...          ...          ...          ...          ...   
3188        -14.248864     1.255221    16.243199    -5.295460   -14.075283   
3189         -0.564849     0.654527     2.408578    21.160658    16.308671   
3190        -12.127022   -18.136812    -5.536861    -7.521840   -14.740608   
3191          1.682288   -11.990126    -1.182592    -2.250385   -18.020168   
3192          6.935294   -12.032179    -8.040059    12.372583     1.533635   

      RNETMIG2015  
0        0.712594  
1       -2.187333  
2       18.293499  
3      -10.543299  
4        1.107861  
...           ...  
3188   -14.070195  
3189     1.520747  
3190   -12.606351  
3191     1.441961  
3192     6.935294  

[3193 rows x 100 columns]

São 100 colunas ao todo, então vamos remover as que não nos interessam e manter apenas os nascimentos de cada ano e estimativa da população. Além disso, vou manter apenas as linhas onde o *sumarrary level* é *40*, pois essas são as linhas que se referem ao census de cada cidade.

In [4]:
colunas = ['STNAME', 'CTYNAME', 'BIRTHS2010', 'BIRTHS2011', 'BIRTHS2012', 'BIRTHS2013', 'BIRTHS2014',
          'BIRTHS2015', 'POPESTIMATE2010', 'POPESTIMATE2011', 'POPESTIMATE2012', 'POPESTIMATE2013', 
          'POPESTIMATE2014', 'POPESTIMATE2015']
df = df[df['SUMLEV'] == 50]
df = df[colunas]
df

STNAME            CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  \
1     Alabama     Autauga County         151         636         615   
2     Alabama     Baldwin County         517        2187        2092   
3     Alabama     Barbour County          70         335         300   
4     Alabama        Bibb County          44         266         245   
5     Alabama      Blount County         183         744         710   
...       ...                ...         ...         ...         ...   
3188  Wyoming  Sweetwater County         167         640         595   
3189  Wyoming       Teton County          76         259         230   
3190  Wyoming       Uinta County          73         324         311   
3191  Wyoming    Washakie County          26         108          90   
3192  Wyoming      Weston County          26          81          74   

      BIRTHS2013  BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  \
1            574         623         600            54660            55253   
2           2160        2186        2240           183193           186659   
3            283         260         269            27341            27226   
4            259         247         253            22861            22733   
5            646         618         603            57373            57711   
...          ...         ...         ...              ...              ...   
3188         657         629         620            43593            44041   
3189         261         249         269            21297            21482   
3190         316         316         316            21102            20912   
3191          95          96          90             8545             8469   
3192          93          77          79             7181             7114   

      POPESTIMATE2012  POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1               55175            55038            55290            55347  
2              190396           195126           199713           203709  
3               27159            26973            26815            26489  
4               22642            22512            22549            22583  
5               57776            57734            57658            57673  
...               ...              ...              ...              ...  
3188            45104            45162            44925            44626  
3189            21697            22347            22905            23125  
3190            20989            21022            20903            20822  
3191             8443             8443             8316             8328  
3192             7065             7160             7185             7234  

[3142 rows x 14 columns]

# Formas de agrupar dados

Vamos agora entender como agrupar esses dados **por estado**. Suponha por exemplo, que queremos saber a média da população em cada estado, ou então a soma, ou mesmo os valores máximos e mínimos de cada cidade por estado.

Para começar a fazer essas análises, precisamos, antes de mais nada, agrupar essas 3142 linhas de acordo com os estados.

In [5]:
df['STNAME'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

Aqui vemos que a coluna *STNAME* possui todos os estados dos EUA, mas que várias linhas são referente ao Alabama, várias linhas são referentes à Florida e assim por diante. A primeira forma que veremos de agrupar é justamente colocar no mesmo grupo aquelas linhas que têm o mesmo valor para uma determinada coluna.

In [6]:
# Utilizamos a função groupby passando o nome da coluna que queremos usar para agrupar
g = df.groupby('STNAME')

In [7]:
# Se tentarmos mostrar a variável g, vemos que ela é um objeto do tipo pandas.core.groupby.generic.DataFrameGroupBy
# Mas ainda não conseguimos entender o que realmente aconteceu com o nosso DataFrame
g

A variável `g` é um **objeto** como vários outro do *pandas* (como uma `Series` ou um `DataFrame`). Isso significa que temos várias funções e atributos dentro desse objeto que nós podemos usar.

Por exemplo, o método describe vai nos gerar várias informações, como em um DataFrame normal, mas de forma agrupada por estado.

In [8]:
g.describe()

BIRTHS2010                                             \
                          count         mean          std     min      25%   
STNAME                                                                       
Alabama                    67.0   212.328358   329.012866    21.0    43.00   
Alaska                     29.0    99.724138   229.620789     2.0     6.00   
Arizona                    15.0  1395.800000  3321.085498    18.0   173.00   
Arkansas                   75.0   125.853333   211.864228    14.0    34.00   
California                 58.0  2126.275862  4627.281089     2.0   112.75   
Colorado                   64.0   262.359375   561.172366     0.0    12.00   
Connecticut                 8.0  1186.125000  1081.829329   290.0   351.25   
Delaware                    3.0   949.000000   655.582947   569.0   570.50   
District of Columbia        1.0  2243.000000          NaN  2243.0  2243.00   
Florida                    67.0   759.283582  1293.232183    17.0    82.00   
Georgia                   159.0   204.333333   453.835199     3.0    34.50   
Hawaii                      5.0   929.200000  1390.219839     0.0   195.00   
Idaho                      44.0   135.704545   240.455922     3.0    23.75   
Illinois                  102.0   410.156863  1874.652468     3.0    40.00   
Indiana                    92.0   227.836957   440.154575    13.0    62.75   
Iowa                       99.0   100.676768   196.334618    10.0    29.00   
Kansas                    105.0    95.161905   283.043865     2.0     9.00   
Kentucky                  120.0   114.091667   256.711426     4.0    31.75   
Louisiana                  64.0   227.703125   305.861919    12.0    56.75   
Maine                      16.0   215.187500   196.058569    35.0    93.75   
Maryland                   24.0   769.208333  1000.857327    41.0   110.00   
Massachusetts              14.0  1331.928571  1295.417335    42.0   274.25   
Michigan                   83.0   350.626506   813.915015     2.0    47.00   
Minnesota                  87.0   201.149425   503.064690     3.0    32.00   
Mississippi                82.0   114.890244   136.077684     6.0    43.25   
Missouri                  115.0   165.321739   389.951314     5.0    29.50   
Montana                    56.0    54.785714    98.431465     0.0     5.75   
Nebraska                   93.0    70.053763   246.597955     0.0     6.00   
Nevada                     17.0   523.058824  1596.206380     3.0    10.00   
New Hampshire              10.0   326.900000   327.353513    82.0   136.00   
New Jersey                 21.0  1263.285714   812.599603   201.0   564.00   
New Mexico                 33.0   201.484848   383.193254     1.0    30.00   
New York                   62.0   990.548387  1908.851484    10.0   128.50   
North Carolina            100.0   299.140000   514.936399    15.0    59.00   
North Dakota               53.0    44.169811    94.113421     0.0     5.00   
Ohio                       88.0   395.556818   708.732264    37.0   105.75   
Oklahoma                   77.0   170.701299   443.136490     5.0    28.00   
Oregon                     36.0   329.916667   543.761522     3.0    49.25   
Pennsylvania               67.0   538.820896   858.733315     5.0   114.50   
Rhode Island                5.0   575.000000   773.294898   106.0   180.00   
South Carolina             46.0   308.260870   352.506418    18.0    73.50   
South Dakota               66.0    45.939394   103.973056     1.0     8.00   
Tennessee                  95.0   206.978947   447.707612    13.0    46.00   
Texas                     254.0   365.224409  1382.661620     0.0    20.00   
Utah                       29.0   474.206897  1058.217686     4.0    41.00   
Vermont                    14.0   118.142857    97.067666    13.0    75.25   
Virginia                  133.0   193.368421   423.093701     0.0    34.00   
Washington                 39.0   570.358974  1125.700238     3.0    54.50   
West Virginia              55.0    92.80000

Veja que, para cada coluna, temos várias informações. Por exemplo, temos a média, desvio padrão, máximo, mínimo, mediana para a população de 2015:

In [9]:
g.describe()['POPESTIMATE2015']

count           mean           std       min        25%  \
STNAME                                                                          
Alabama                67.0   72522.074627  1.058137e+05    8479.0   18163.00   
Alaska                 29.0   25463.172414  5.867997e+04     613.0    2534.00   
Arizona                15.0  455204.333333  1.056990e+06    9529.0   49810.00   
Arkansas               75.0   39709.386667  6.027020e+04    5229.0   12330.00   
California             58.0  674910.655172  1.477043e+06    1110.0   47048.25   
Colorado               64.0   85258.968750  1.731548e+05     701.0    5761.50   
Connecticut             8.0  448860.750000  3.778265e+05  116573.0  160902.25   
Delaware                3.0  315311.333333  2.101734e+05  173533.0  194577.50   
District of Columbia    1.0  672228.000000           NaN  672228.0  672228.00   
Florida                67.0  302556.298507  4.830767e+05    8331.0   27461.00   
Georgia               159.0   64244.402516  1.386611e+05    1639.0   11382.00   
Hawaii                  5.0  286320.600000  4.056992e+05      89.0   71735.00   
Idaho                  44.0   37612.045455  7.352363e+04     880.0    7500.50   
Illinois              102.0  126078.382353  5.319565e+05    4135.0   14284.25   
Indiana                92.0   71953.043478  1.208524e+05    5938.0   20914.00   
Iowa                   99.0   31554.535354  5.684560e+04    3796.0   10092.00   
Kansas                105.0   27729.914286  7.813176e+04    1330.0    3130.00   
Kentucky              120.0   36875.766667  7.684637e+04    2138.0   12258.75   
Louisiana              64.0   72980.062500  9.864068e+04    4740.0   20030.50   
Maine                  16.0   83083.000000  7.509957e+04   16931.0   34854.00   
Maryland               24.0  250266.708333  3.060913e+05   19787.0   46056.00   
Massachusetts          14.0  485315.857143  4.353452e+05   10925.0  136194.00   
Michigan               83.0  119549.108434  2.589048e+05    2168.0   22519.50   
Minnesota              87.0   63098.781609  1.523805e+05    3401.0   10791.50   
Mississippi            82.0   36491.865854  4.296988e+04    1337.0   12847.00   
Missouri              115.0   52901.495652  1.241784e+05    2057.0   10167.50   
Montana                56.0   18445.517857  3.194864e+04     475.0    2994.50   
Nebraska               93.0   20389.139785  6.663466e+04     456.0    2883.00   
Nevada                 17.0  170049.705882  5.119906e+05     829.0    5036.00   
New Hampshire          10.0  133060.800000  1.246618e+05   31212.0   50624.00   
New Jersey             21.0  426572.047619  2.624544e+05   64180.0  155854.00   
New Mexico             33.0   63185.121212  1.207211e+05     698.0   11282.00   
New York               62.0  319286.951613  5.544434e+05    4712.0   49896.50   
North Carolina        100.0  100428.020000  1.582575e+05    4070.0   24993.25   
North Dakota           53.0   14281.641509  2.870866e+04     767.0    2417.00   
Ohio                   88.0  131970.715909  2.162611e+05   13048.0   36982.25   
Oklahoma               77.0   50796.597403  1.152736e+05    2216.0   10483.00   
Oregon                 36.0  111916.027778  1.744656e+05    1358.0   20223.75   
Pennsylvania           67.0  191082.134328  2.738224e+05    4732.0   43421.00   
Rhode Island            5.0  211259.600000  2.400540e+05   49084.0   82423.00   
South Carolina         46.0  106437.956522  1.187606e+05    9433.0   27564.75   
South Dakota           66.0   13007.106061  2.665543e+04     924.0    2959.75   
Tennessee              95.0   69476.831579  1.313463e+05    5146.0   17845.00   
Texas                 254.0  108146.118110  3.902188e+05     112.0    6929.00   
Utah                   29.0  103307.551724  2.294843e+05    1109.0    9516.00   
Vermont                14.0   44717.285714  3.757316e+04    6163.0   27549.75   
Virginia              133.0   63030.022556  1.249163e+05    2214.0   14785.00   
Washington             39.0  183855.153846  3.734423e

**Importante:** note que, depois de aplicarmos a função `describe`, o resultado é um `DataFrame`. Ou seja, podemos usar tudo aquilo que já sabemos sobre `DataFrame`'s (como acessar uma coluna, igual eu fiz acima).

In [10]:
type(g.describe())

pandas.core.frame.DataFrame

O `describe` nos traz muitas informações, mas podemos também aplicar outras funções para ter apenas aquilo que desejamos (e de forma mais rápida):

In [11]:
g.mean().head()

BIRTHS2010   BIRTHS2011   BIRTHS2012   BIRTHS2013   BIRTHS2014  \
STNAME                                                                        
Alabama      212.328358   890.880597   881.522388   864.746269   870.656716   
Alaska        99.724138   403.827586   383.655172   391.310345   395.241379   
Arizona     1395.800000  5740.266667  5704.400000  5740.200000  5770.400000   
Arkansas     125.853333   512.600000   514.720000   506.386667   505.493333   
California  2126.275862  8789.086207  8576.396552  8614.086207  8613.500000   

             BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
STNAME                                                                       
Alabama      870.223881     71420.313433     71658.328358     71881.925373   
Alaska       396.482759     24621.413793     24921.379310     25214.758621   
Arizona     5825.666667    427213.866667    431248.800000    436884.133333   
Arkansas     504.266667     38965.253333     39180.506667     39326.653333   
California  8687.051724    643691.017241    650000.586207    656138.879310   

            POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
STNAME                                                         
Alabama        72097.507463     72334.492537     72522.074627  
Alaska         25429.034483     25415.379310     25463.172414  
Arizona       442053.266667    448585.533333    455204.333333  
Arkansas       39439.426667     39557.800000     39709.386667  
California    662312.551724    668832.603448    674910.655172

In [12]:
g.max().head()

CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  \
STNAME                                                                      
Alabama                Winston County        2104        8831        9007   
Alaska      Yukon-Koyukuk Census Area        1162        4783        4591   
Arizona                   Yuma County       13123       53377       54292   
Arkansas                  Yell County        1425        5800        5654   
California                Yuba County       31740      132649      129157   

            BIRTHS2013  BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME                                                            
Alabama           8843        8621        8639           658116   
Alaska            4698        4810        4796           293405   
Arizona          54627       55060       55671          3825597   
Arkansas          5656        5497        5514           383628   
California      130647      129468      130070          9826009   

            POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME                                                          
Alabama              657779           657704           658913   
Alaska               296333           298613           301537   
Arizona             3871957          3945460          4015328   
Arkansas             386827           388984           391352   
California          9896602          9970436         10045175   

            POPESTIMATE2014  POPESTIMATE2015  
STNAME                                        
Alabama              660368           660367  
Alaska               300357           298695  
Arizona             4090022          4167947  
Arkansas             392490           392664  
California         10109436         10170292

In [13]:
g.count().head()

CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME                                                                
Alabama          67          67          67          67          67   
Alaska           29          29          29          29          29   
Arizona          15          15          15          15          15   
Arkansas         75          75          75          75          75   
California       58          58          58          58          58   

            BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  \
STNAME                                                                 
Alabama             67          67               67               67   
Alaska              29          29               29               29   
Arizona             15          15               15               15   
Arkansas            75          75               75               75   
California          58          58               58               58   

            POPESTIMATE2012  POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
STNAME                                                                          
Alabama                  67               67               67               67  
Alaska                   29               29               29               29  
Arizona                  15               15               15               15  
Arkansas                 75               75               75               75  
California               58               58               58               58

A lista completa de operações pode ser encontrada [aqui](https://pandas.pydata.org/pandas-docs/stable/reference/groupby.html#computations-descriptive-stats).

# Acessando grupos

Em muitos casos, nós precisamos acessar um grupo específico ou mesmo iterar sobre nossos grupos. 

Para acessar um grupo específico, usamos a função `get_group` e passamos o grupo desejado.

In [14]:
g.get_group('California')

STNAME                 CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  \
191  California          Alameda County        4877       19219       18972   
192  California           Alpine County           2           6           6   
193  California           Amador County          78         264         280   
194  California            Butte County         623        2440        2397   
195  California        Calaveras County          74         333         356   
196  California           Colusa County          70         336         302   
197  California     Contra Costa County        3011       12335       11941   
198  California        Del Norte County          89         336         323   
199  California        El Dorado County         382        1631        1598   
200  California           Fresno County        3893       16321       15950   
201  California            Glenn County         107         426         374   
202  California         Humboldt County         384        1546        1434   
203  California         Imperial County         687        3128        3022   
204  California             Inyo County          39         194         230   
205  California             Kern County        3487       14405       14206   
206  California            Kings County         569        2580        2397   
207  California             Lake County         162         743         726   
208  California           Lassen County          85         299         303   
209  California      Los Angeles County       31740      132649      129157   
210  California           Madera County         574        2440        2375   
211  California            Marin County         588        2384        2327   
212  California         Mariposa County          31         148         142   
213  California        Mendocino County         305        1010        1122   
214  California           Merced County        1066        4302        4304   
215  California            Modoc County          27         100          76   
216  California             Mono County          31         153         145   
217  California         Monterey County        1673        6863        6808   
218  California             Napa County         388        1501        1529   
219  California           Nevada County         215         771         771   
220  California           Orange County        9304       38238       37842   
221  California           Placer County         953        3851        3692   
222  California           Plumas County          41         168         157   
223  California        Riverside County        7219       30999       30094   
224  California       Sacramento County        5026       20194       19537   
225  California       San Benito County         190         763         716   
226  California   San Bernardino County        7465       31432       30164   
227  California        San Diego County       10719       44694       43841   
228  California    San Francisco County        2121        8899        8782   
229  California      San Joaquin County        2548       10550       10085   
230  California  San Luis Obispo County         703        2644        2626   
231  California        San Mateo County        2285        9141        9012   
232  California    Santa Barbara County        1510        5749        5639   
233  California      Santa Clara County        5866       24021       23416   
234  California       Santa Cruz County         811        3245        3163   
235  California           Shasta County         533        2107        2058   
236  California           Sierra County           7          17          21   
237  California         Siskiyou County          90         462         494   
238  California           Solano County        1192        5174        5072   
239  California           Sonoma County        1421        5310        5045   
240  California       Stanislaus County        1923        

Note novamente que o resultado é um `DataFrame` onde todos os valores da coluna *STNAME* são 'California'.

O resultado é exatamente o mesmo que quando fazemos:

In [15]:
df[df['STNAME'] == 'California']

STNAME                 CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  \
191  California          Alameda County        4877       19219       18972   
192  California           Alpine County           2           6           6   
193  California           Amador County          78         264         280   
194  California            Butte County         623        2440        2397   
195  California        Calaveras County          74         333         356   
196  California           Colusa County          70         336         302   
197  California     Contra Costa County        3011       12335       11941   
198  California        Del Norte County          89         336         323   
199  California        El Dorado County         382        1631        1598   
200  California           Fresno County        3893       16321       15950   
201  California            Glenn County         107         426         374   
202  California         Humboldt County         384        1546        1434   
203  California         Imperial County         687        3128        3022   
204  California             Inyo County          39         194         230   
205  California             Kern County        3487       14405       14206   
206  California            Kings County         569        2580        2397   
207  California             Lake County         162         743         726   
208  California           Lassen County          85         299         303   
209  California      Los Angeles County       31740      132649      129157   
210  California           Madera County         574        2440        2375   
211  California            Marin County         588        2384        2327   
212  California         Mariposa County          31         148         142   
213  California        Mendocino County         305        1010        1122   
214  California           Merced County        1066        4302        4304   
215  California            Modoc County          27         100          76   
216  California             Mono County          31         153         145   
217  California         Monterey County        1673        6863        6808   
218  California             Napa County         388        1501        1529   
219  California           Nevada County         215         771         771   
220  California           Orange County        9304       38238       37842   
221  California           Placer County         953        3851        3692   
222  California           Plumas County          41         168         157   
223  California        Riverside County        7219       30999       30094   
224  California       Sacramento County        5026       20194       19537   
225  California       San Benito County         190         763         716   
226  California   San Bernardino County        7465       31432       30164   
227  California        San Diego County       10719       44694       43841   
228  California    San Francisco County        2121        8899        8782   
229  California      San Joaquin County        2548       10550       10085   
230  California  San Luis Obispo County         703        2644        2626   
231  California        San Mateo County        2285        9141        9012   
232  California    Santa Barbara County        1510        5749        5639   
233  California      Santa Clara County        5866       24021       23416   
234  California       Santa Cruz County         811        3245        3163   
235  California           Shasta County         533        2107        2058   
236  California           Sierra County           7          17          21   
237  California         Siskiyou County          90         462         494   
238  California           Solano County        1192        5174        5072   
239  California           Sonoma County        1421        5310        5045   
240  California       Stanislaus County        1923        

A diferença é que `g` possui muitos outros grupos e é muito mais prático, como já vimos.

Para iterar pelos grupos, podemos usar um `for`.

In [16]:
# Para iterar é como se fosse um dicionário com a chave do grupo e o DataFrame
for chave, grupo in g:
    print(f'O nome do grupo é {chave} e o shape é {grupo.shape}.')

O nome do grupo é Alabama e o shape é (67, 14).
O nome do grupo é Alaska e o shape é (29, 14).
O nome do grupo é Arizona e o shape é (15, 14).
O nome do grupo é Arkansas e o shape é (75, 14).
O nome do grupo é California e o shape é (58, 14).
O nome do grupo é Colorado e o shape é (64, 14).
O nome do grupo é Connecticut e o shape é (8, 14).
O nome do grupo é Delaware e o shape é (3, 14).
O nome do grupo é District of Columbia e o shape é (1, 14).
O nome do grupo é Florida e o shape é (67, 14).
O nome do grupo é Georgia e o shape é (159, 14).
O nome do grupo é Hawaii e o shape é (5, 14).
O nome do grupo é Idaho e o shape é (44, 14).
O nome do grupo é Illinois e o shape é (102, 14).
O nome do grupo é Indiana e o shape é (92, 14).
O nome do grupo é Iowa e o shape é (99, 14).
O nome do grupo é Kansas e o shape é (105, 14).
O nome do grupo é Kentucky e o shape é (120, 14).
O nome do grupo é Louisiana e o shape é (64, 14).
O nome do grupo é Maine e o shape é (16, 14).
O nome do grupo é Maryl

Espero que agora esteja começando a ficar claro quais são as funcionalidade desse tipo de objeto.

Para terminar essa primeira parte, vale a pena dizer que podemos indexar colunas de um objeto `groupby`:

In [17]:
g[['BIRTHS2010','BIRTHS2011','BIRTHS2012','POPESTIMATE2014','POPESTIMATE2015']].mean().head()

BIRTHS2010   BIRTHS2011   BIRTHS2012  POPESTIMATE2014  \
STNAME                                                               
Alabama      212.328358   890.880597   881.522388     72334.492537   
Alaska        99.724138   403.827586   383.655172     25415.379310   
Arizona     1395.800000  5740.266667  5704.400000    448585.533333   
Arkansas     125.853333   512.600000   514.720000     39557.800000   
California  2126.275862  8789.086207  8576.396552    668832.603448   

            POPESTIMATE2015  
STNAME                       
Alabama        72522.074627  
Alaska         25463.172414  
Arizona       455204.333333  
Arkansas       39709.386667  
California    674910.655172

# Outras formas de se agrupar

Agora que já entendemos alguns usos básicos de um agrupamento. Vamos voltar um pouco para a função `groupby` e ver rapidamente outras formas de utiliza-la.

### Várias colunas
A primeira forma é passando um array de colunas ao invés de uma única coluna.

In [18]:
df1 = pd.DataFrame({'col1' : [1, 1, 2, 2, 2, 3],
                  'col2' : ['no', 'yes', 'no', 'no', 'yes', 'no'],
                  'col_val1' : np.random.rand(6),
                  'col_val2' : np.random.randn(6)*10})
df1

col1 col2  col_val1   col_val2
0     1   no  0.464668  -4.645701
1     1  yes  0.128543 -11.909457
2     2   no  0.267201   3.637542
3     2   no  0.984900   1.515328
4     2  yes  0.192741   4.025386
5     3   no  0.437094  -7.741972

In [19]:
df1.groupby(['col1', 'col2']).mean()

col_val1   col_val2
col1 col2                     
1    no    0.464668  -4.645701
     yes   0.128543 -11.909457
2    no    0.626050   2.576435
     yes   0.192741   4.025386
3    no    0.437094  -7.741972

Perceba que agora temos um DataFrame com um MultiIndex onde o nível superior é 'col1' e o inferior é 'col2'. E note que não há uma linha `(3, 'yes')` porque essa combinação não existe.

### Por nível em um MultiIndex
Falando em MultiIndex, outra forma de agrupar é informando o nível em um MultiIndex.

In [20]:
colunas = pd.MultiIndex.from_arrays([['US']*3 +['JP']*2,
                                     [1, 3, 5, 1, 3]],
                                    names=['cty', 'num'])
hier_df = pd.DataFrame(np.random.randn(4, 5), columns=colunas)
hier_df

cty        US                            JP          
num         1         3         5         1         3
0   -1.216321  1.246749 -1.184661 -0.459296 -0.653470
1   -2.205788 -0.151072 -2.231622 -0.178378 -0.461127
2    0.133232  1.159642 -0.259300 -0.675201 -0.660947
3    1.710449  0.702046  0.955168  0.253405 -0.090389

In [21]:
hier_df.groupby(level='cty', axis='columns').max()

cty        JP        US
0   -0.459296  1.246749
1   -0.178378 -0.151072
2   -0.660947  1.159642
3    0.253405  1.710449

### Usando uma função

Podemos também agrupar utilizando uma função (criada por nós ou não). 

A função vai receber o índice de cada linha e deve devolver um valor que pode ser comparável. Então vamos criar os grupos de acordo com os resultados iguais da função.

In [22]:
df2 = pd.DataFrame({'nome' : ['João', 'Cláudia', 'Mariana', 'Mario', 'Joana', 'Lucas', 'Marcos'],
                   'nota1' : np.random.randint(0, 11, size=7),
                   'nota2' : np.random.randint(0, 11, size=7)})
df2 = df2.set_index('nome') # É importante que o índice seja os nomes, pois a função vai receber esses nomes
df2

nota1  nota2
nome                 
João         7      6
Cláudia      5      1
Mariana      1      5
Mario        3      4
Joana        3      4
Lucas        1      8
Marcos       4      6

In [23]:
# Vamos criar uma função que pega uma string e pega suas primeiras duas letras
def duas_letras(s):
    # Sempre que tiver dúvida quanto a o que a função está recebendo, experimente printar o valor ou seu tipo
    #print(s)
    #print(type(s))
    return s[:2]
df2.groupby(duas_letras).max()

nota1  nota2
Cl      5      1
Jo      7      6
Lu      1      8
Ma      4      6

Veja como nós agrupamos as pessoas de acordo com as duas primeiras letras de seu nome.

Poderíamos usar também funções do Python, como a função `len`:

In [24]:
df2.groupby(len).max()

nota1  nota2
4      7      6
5      3      8
6      4      6
7      5      5

Se quisermos ver quem percente a que grupo, uma forma prática é utilizando o *atributo* `groups` do nosso objeto `groupby`.

In [25]:
df2.groupby(len).groups

{4: ['João'], 5: ['Mario', 'Joana', 'Lucas'], 6: ['Marcos'], 7: ['Cláudia', 'Mariana']}

# Agregações

Vamos agora passar para as principais aplicações e utilizades dos agrupamentos:
- Gerar agregações
- Gerar transformações
- Filtrar grupos

Começando com as agregações.

Sempre que temos uma lista e geramos um número com base naquela lista, estamos gerando uma agregação. Ou seja, uma agregação é sempre uma operação que transforma uma lista de valores em um único valor. Exemplos de agregação são soma, média, mediana, desvio padrão, máximo, mínimo, produto, entre outras agregações que nós mesmos podemos criar (agregações não se restringem a números).

Quando se trata de grupos, as agregações vão ser feitas dentro de cada grupo. Ou seja, se utilizamos uma média, vamos ter a média dos valores de cada grupo.

No pandas, as funções `agg` ou `aggregate` são utilizadas para fazer agregações (não existe diferença nenhuma entre elas, `agg` é um *alias* para *aggregate*).

In [26]:
# Vamos voltar ao nosso banco de dados do censo
g.agg(np.mean).head(10)

BIRTHS2010   BIRTHS2011   BIRTHS2012   BIRTHS2013  \
STNAME                                                                     
Alabama                212.328358   890.880597   881.522388   864.746269   
Alaska                  99.724138   403.827586   383.655172   391.310345   
Arizona               1395.800000  5740.266667  5704.400000  5740.200000   
Arkansas               125.853333   512.600000   514.720000   506.386667   
California            2126.275862  8789.086207  8576.396552  8614.086207   
Colorado               262.359375  1029.937500  1011.562500  1011.546875   
Connecticut           1186.125000  4704.500000  4606.125000  4516.750000   
Delaware               949.000000  3766.333333  3719.333333  3630.000000   
District of Columbia  2243.000000  9196.000000  9234.000000  9447.000000   
Florida                759.283582  3182.343284  3194.567164  3179.164179   

                       BIRTHS2014   BIRTHS2015  POPESTIMATE2010  \
STNAME                                                            
Alabama                870.656716   870.223881     71420.313433   
Alaska                 395.241379   396.482759     24621.413793   
Arizona               5770.400000  5825.666667    427213.866667   
Arkansas               505.493333   504.266667     38965.253333   
California            8613.500000  8687.051724    643691.017241   
Colorado              1033.734375  1047.765625     78878.968750   
Connecticut           4522.750000  4528.125000    447464.625000   
Delaware              3642.333333  3677.666667    299930.333333   
District of Columbia  9415.000000  9593.000000    605126.000000   
Florida               3255.044776  3292.955224    281341.641791   

                      POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME                                                                    
Alabama                  71658.328358     71881.925373     72097.507463   
Alaska                   24921.379310     25214.758621     25429.034483   
Arizona                 431248.800000    436884.133333    442053.266667   
Arkansas                 39180.506667     39326.653333     39439.426667   
California              650000.586207    656138.879310    662312.551724   
Colorado                 79991.875000     81120.796875     82361.437500   
Connecticut             448719.875000    449192.625000    449646.000000   
Delaware                302638.666667    305699.666667    308451.000000   
District of Columbia    620472.000000    635342.000000    649540.000000   
Florida                 285157.208955    288836.134328    292454.731343   

                      POPESTIMATE2014  POPESTIMATE2015  
STNAME                                                  
Alabama                  72334.492537     72522.074627  
Alaska                   25415.379310     25463.172414  
Arizona                 448585.533333    455204.333333  
Arkansas                 39557.800000     39709.386667  
California              668832.603448    674910.655172  
Colorado                 83681.062500     85258.968750  
Connecticut             449345.250000    448860.750000  
Delaware                311989.333333    315311.333333  
District of Columbia    659836.000000    672228.000000  
Florida                 297098.044776    302556.298507

Note que isso é exatamente o que estávamos fazendo até agora:

In [27]:
g.mean().head(10)

BIRTHS2010   BIRTHS2011   BIRTHS2012   BIRTHS2013  \
STNAME                                                                     
Alabama                212.328358   890.880597   881.522388   864.746269   
Alaska                  99.724138   403.827586   383.655172   391.310345   
Arizona               1395.800000  5740.266667  5704.400000  5740.200000   
Arkansas               125.853333   512.600000   514.720000   506.386667   
California            2126.275862  8789.086207  8576.396552  8614.086207   
Colorado               262.359375  1029.937500  1011.562500  1011.546875   
Connecticut           1186.125000  4704.500000  4606.125000  4516.750000   
Delaware               949.000000  3766.333333  3719.333333  3630.000000   
District of Columbia  2243.000000  9196.000000  9234.000000  9447.000000   
Florida                759.283582  3182.343284  3194.567164  3179.164179   

                       BIRTHS2014   BIRTHS2015  POPESTIMATE2010  \
STNAME                                                            
Alabama                870.656716   870.223881     71420.313433   
Alaska                 395.241379   396.482759     24621.413793   
Arizona               5770.400000  5825.666667    427213.866667   
Arkansas               505.493333   504.266667     38965.253333   
California            8613.500000  8687.051724    643691.017241   
Colorado              1033.734375  1047.765625     78878.968750   
Connecticut           4522.750000  4528.125000    447464.625000   
Delaware              3642.333333  3677.666667    299930.333333   
District of Columbia  9415.000000  9593.000000    605126.000000   
Florida               3255.044776  3292.955224    281341.641791   

                      POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME                                                                    
Alabama                  71658.328358     71881.925373     72097.507463   
Alaska                   24921.379310     25214.758621     25429.034483   
Arizona                 431248.800000    436884.133333    442053.266667   
Arkansas                 39180.506667     39326.653333     39439.426667   
California              650000.586207    656138.879310    662312.551724   
Colorado                 79991.875000     81120.796875     82361.437500   
Connecticut             448719.875000    449192.625000    449646.000000   
Delaware                302638.666667    305699.666667    308451.000000   
District of Columbia    620472.000000    635342.000000    649540.000000   
Florida                 285157.208955    288836.134328    292454.731343   

                      POPESTIMATE2014  POPESTIMATE2015  
STNAME                                                  
Alabama                  72334.492537     72522.074627  
Alaska                   25415.379310     25463.172414  
Arizona                 448585.533333    455204.333333  
Arkansas                 39557.800000     39709.386667  
California              668832.603448    674910.655172  
Colorado                 83681.062500     85258.968750  
Connecticut             449345.250000    448860.750000  
Delaware                311989.333333    315311.333333  
District of Columbia    659836.000000    672228.000000  
Florida                 297098.044776    302556.298507

Mas então qual a diferença? Fazer apenas `.mean()` parece muito mais fácil.

A diferença é que a função `agg` nos permite fazer muitos mais. E é isso que vamos ver agora:

### Várias agregações
Com `agg` podemos passar quantas funções de agregação nós quisermos e vamos gerar um DataFrame com colunas hierarquicas com os nomes dessas funções:

In [28]:
g.agg([np.mean, np.sum]).head(10)

BIRTHS2010           BIRTHS2011           BIRTHS2012  \
                             mean     sum         mean     sum         mean   
STNAME                                                                        
Alabama                212.328358   14226   890.880597   59689   881.522388   
Alaska                  99.724138    2892   403.827586   11711   383.655172   
Arizona               1395.800000   20937  5740.266667   86104  5704.400000   
Arkansas               125.853333    9439   512.600000   38445   514.720000   
California            2126.275862  123324  8789.086207  509767  8576.396552   
Colorado               262.359375   16791  1029.937500   65916  1011.562500   
Connecticut           1186.125000    9489  4704.500000   37636  4606.125000   
Delaware               949.000000    2847  3766.333333   11299  3719.333333   
District of Columbia  2243.000000    2243  9196.000000    9196  9234.000000   
Florida                759.283582   50872  3182.343284  213217  3194.567164   

                               BIRTHS2013           BIRTHS2014          ...  \
                         sum         mean     sum         mean     sum  ...   
STNAME                                                                  ...   
Alabama                59062   864.746269   57938   870.656716   58334  ...   
Alaska                 11126   391.310345   11348   395.241379   11462  ...   
Arizona                85566  5740.200000   86103  5770.400000   86556  ...   
Arkansas               38604   506.386667   37979   505.493333   37912  ...   
California            497431  8614.086207  499617  8613.500000  499583  ...   
Colorado               64740  1011.546875   64739  1033.734375   66159  ...   
Connecticut            36849  4516.750000   36134  4522.750000   36182  ...   
Delaware               11158  3630.000000   10890  3642.333333   10927  ...   
District of Columbia    9234  9447.000000    9447  9415.000000    9415  ...   
Florida               214036  3179.164179  213004  3255.044776  218088  ...   

                     POPESTIMATE2011           POPESTIMATE2012            \
                                mean       sum            mean       sum   
STNAME                                                                     
Alabama                 71658.328358   4801108    71881.925373   4816089   
Alaska                  24921.379310    722720    25214.758621    731228   
Arizona                431248.800000   6468732   436884.133333   6553262   
Arkansas                39180.506667   2938538    39326.653333   2949499   
California             650000.586207  37700034   656138.879310  38056055   
Colorado                79991.875000   5119480    81120.796875   5191731   
Connecticut            448719.875000   3589759   449192.625000   3593541   
Delaware               302638.666667    907916   305699.666667    917099   
District of Columbia   620472.000000    620472   635342.000000    635342   
Florida                285157.208955  19105533   288836.134328  19352021   

                     POPESTIMATE2013           POPESTIMATE2014            \
                                mean       sum            mean       sum   
STNAME                                                                     
Alabama                 72097.507463   4830533    72334.492537   4846411   
Alaska                  25429.034483    737442    25415.379310    737046   
Arizona                442053.266667   6630799   448585.533333   6728783   
Arkansas                39439.426667   2957957    39557.800000   2966835   
California             662312.551724  38414128   668832.603448  38792291   
Colorado                82361.437500   5271132    83681.062500   5355588   
Connecticut            449646.000000   3597168   449345.250000   3594762   
Delaware               308451.000000    925353   311989.333333    935968   
District of Columbia   649540.000000    649540   659836.000000    659836   
Florida                292454.731343  19594467   297098.044776  19

### Agregações customizáveis
Além disso, podemos passar nossas próprias funções para a `agg`:

In [29]:
# O exemplo é besta, mas em muitos casos nós precisamos definir nossas próprias funções para gerar agregações 
# exatamente como desejamos

def meia_media(s):
    return s.mean() / 2
def soma_mais_10(s):
    return s.sum() + 10

g.agg([meia_media, soma_mais_10]).head(10)

BIRTHS2010                BIRTHS2011               \
                       meia_media soma_mais_10   meia_media soma_mais_10   
STNAME                                                                     
Alabama                106.164179        14236   445.440299        59699   
Alaska                  49.862069         2902   201.913793        11721   
Arizona                697.900000        20947  2870.133333        86114   
Arkansas                62.926667         9449   256.300000        38455   
California            1063.137931       123334  4394.543103       509777   
Colorado               131.179688        16801   514.968750        65926   
Connecticut            593.062500         9499  2352.250000        37646   
Delaware               474.500000         2857  1883.166667        11309   
District of Columbia  1121.500000         2253  4598.000000         9206   
Florida                379.641791        50882  1591.171642       213227   

                       BIRTHS2012                BIRTHS2013               \
                       meia_media soma_mais_10   meia_media soma_mais_10   
STNAME                                                                     
Alabama                440.761194        59072   432.373134        57948   
Alaska                 191.827586        11136   195.655172        11358   
Arizona               2852.200000        85576  2870.100000        86113   
Arkansas               257.360000        38614   253.193333        37989   
California            4288.198276       497441  4307.043103       499627   
Colorado               505.781250        64750   505.773438        64749   
Connecticut           2303.062500        36859  2258.375000        36144   
Delaware              1859.666667        11168  1815.000000        10900   
District of Columbia  4617.000000         9244  4723.500000         9457   
Florida               1597.283582       214046  1589.582090       213014   

                       BIRTHS2014               ... POPESTIMATE2011  \
                       meia_media soma_mais_10  ...      meia_media   
STNAME                                          ...                   
Alabama                435.328358        58344  ...    35829.164179   
Alaska                 197.620690        11472  ...    12460.689655   
Arizona               2885.200000        86566  ...   215624.400000   
Arkansas               252.746667        37922  ...    19590.253333   
California            4306.750000       499593  ...   325000.293103   
Colorado               516.867188        66169  ...    39995.937500   
Connecticut           2261.375000        36192  ...   224359.937500   
Delaware              1821.166667        10937  ...   151319.333333   
District of Columbia  4707.500000         9425  ...   310236.000000   
Florida               1627.522388       218098  ...   142578.604478   

                                  POPESTIMATE2012               \
                     soma_mais_10      meia_media soma_mais_10   
STNAME                                                           
Alabama                   4801118    35940.962687      4816099   
Alaska                     722730    12607.379310       731238   
Arizona                   6468742   218442.066667      6553272   
Arkansas                  2938548    19663.326667      2949509   
California               37700044   328069.439655     38056065   
Colorado                  5119490    40560.398438      5191741   
Connecticut               3589769   224596.312500      3593551   
Delaware                   907926   152849.833333       917109   
District of Columbia       620482   317671.000000       635352   
Florida                  19105543   144418.067164     19352031   

                     POPESTIMATE2013              POPESTIMATE2014  \
                          meia_media soma_mais_10      meia_media   
STNAME                                                              
Alabama                 36048.753731      4830543    36

Note que a função passada para `agg` recebe como parâmetro uma `Series` que corresponde a uma coluna do agrupamento. Vamos provar isso com `DataFrame` menor:

In [30]:
df2

nota1  nota2
nome                 
João         7      6
Cláudia      5      1
Mariana      1      5
Mario        3      4
Joana        3      4
Lucas        1      8
Marcos       4      6

In [31]:
def verifica_tipo(s):
    print(type(s))
    print(s)
    print('-' * 100, end='\n\n')
    if s.shape[0] > 1:
        return 1000
    else:
        return 0
    
df2.groupby(duas_letras).agg(verifica_tipo)

<class 'pandas.core.series.Series'>
Cláudia    5
Name: nota1, dtype: int64
----------------------------------------------------------------------------------------------------

<class 'pandas.core.series.Series'>
João     7
Joana    3
Name: nota1, dtype: int64
----------------------------------------------------------------------------------------------------

<class 'pandas.core.series.Series'>
Lucas    1
Name: nota1, dtype: int64
----------------------------------------------------------------------------------------------------

<class 'pandas.core.series.Series'>
Mariana    1
Mario      3
Marcos     4
Name: nota1, dtype: int64
----------------------------------------------------------------------------------------------------

<class 'pandas.core.series.Series'>
Cláudia    1
Name: nota2, dtype: int64
----------------------------------------------------------------------------------------------------

<class 'pandas.core.series.Series'>
João     6
Joana    4
Name: nota2, dtype: int6

nota1  nota2
Cl      0      0
Jo   1000   1000
Lu      0      0
Ma   1000   1000

Veja como cada grupo aparece duas vezes, um para cada coluna do `DataFrame` original.

### Nomear agregações
Podemos, agora, querer renomear os nomes que aparecem em cada coluna da tabela agregada. Veja como podemos ter alguns casos em que renomar as colunas é bastante desejável:

In [32]:
def uma_funcao_com_um_nome_desnecessariamente_grande(s):
    return s.mean() / 2

g.agg([uma_funcao_com_um_nome_desnecessariamente_grande,
       lambda x : x.sum(),
       lambda x : x.max() - x.min()]).head(10)

BIRTHS2010  \
                     uma_funcao_com_um_nome_desnecessariamente_grande   
STNAME                                                                  
Alabama                                                    106.164179   
Alaska                                                      49.862069   
Arizona                                                    697.900000   
Arkansas                                                    62.926667   
California                                                1063.137931   
Colorado                                                   131.179688   
Connecticut                                                593.062500   
Delaware                                                   474.500000   
District of Columbia                                      1121.500000   
Florida                                                    379.641791   

                                            \
                     <lambda_0> <lambda_1>   
STNAME                                       
Alabama                   14226       2083   
Alaska                     2892       1160   
Arizona                   20937      13105   
Arkansas                   9439       1411   
California               123324      31738   
Colorado                  16791       2405   
Connecticut                9489       2429   
Delaware                   2847       1137   
District of Columbia       2243          0   
Florida                   50872       7258   

                                                           BIRTHS2011  \
                     uma_funcao_com_um_nome_desnecessariamente_grande   
STNAME                                                                  
Alabama                                                    445.440299   
Alaska                                                     201.913793   
Arizona                                                   2870.133333   
Arkansas                                                   256.300000   
California                                                4394.543103   
Colorado                                                   514.968750   
Connecticut                                               2352.250000   
Delaware                                                  1883.166667   
District of Columbia                                      4598.000000   
Florida                                                   1591.171642   

                                            \
                     <lambda_0> <lambda_1>   
STNAME                                       
Alabama                   59689       8736   
Alaska                    11711       4775   
Arizona                   86104      53263   
Arkansas                  38445       5759   
California               509767     132643   
Colorado                  65916       9584   
Connecticut               37636       9224   
Delaware                  11299       4673   
District of Columbia       9196          0   
Florida                  213217      31247   

                                                           BIRTHS2012  \
                     uma_funcao_com_um_nome_desnecessariamente_grande   
STNAME                                                                  
Alabama                                                    440.761194   
Alaska                                                     191.827586   
Arizona                                                   2852.200000   
Arkansas                                                   257.360000   
California                                                4288.198276   
Colorado                                                   505.781250   
Connecticut                                               2303.062500   
Delaware                                                  1859.666667   
District of Columbia                                      4617.000000   
Florida                                                   1597.283582   

           

Veja que temos uma função com um nome gigante e duas funções lambda (funções sem nome) que acabam não dando um resultado muito desejável. 

Podemos, entretanto, passar nomes para nossas colunas através da função `agg`. Ao invés de passar uma lista de funções, passamos uma lista de tuplas onde o primeiro elemento é o nome e o segundo a função:

In [33]:
g.agg([('meia_soma', uma_funcao_com_um_nome_desnecessariamente_grande),
       ('soma', lambda x : x.sum()),
       ('amplitude', lambda x : x.max() - x.min())]).head(10)

BIRTHS2010                     BIRTHS2011          \
                        meia_soma    soma amplitude    meia_soma    soma   
STNAME                                                                     
Alabama                106.164179   14226      2083   445.440299   59689   
Alaska                  49.862069    2892      1160   201.913793   11711   
Arizona                697.900000   20937     13105  2870.133333   86104   
Arkansas                62.926667    9439      1411   256.300000   38445   
California            1063.137931  123324     31738  4394.543103  509767   
Colorado               131.179688   16791      2405   514.968750   65916   
Connecticut            593.062500    9489      2429  2352.250000   37636   
Delaware               474.500000    2847      1137  1883.166667   11299   
District of Columbia  1121.500000    2243         0  4598.000000    9196   
Florida                379.641791   50872      7258  1591.171642  213217   

                                 BIRTHS2012                     BIRTHS2013  \
                     amplitude    meia_soma    soma amplitude    meia_soma   
STNAME                                                                       
Alabama                   8736   440.761194   59062      8902   432.373134   
Alaska                    4775   191.827586   11126      4584   195.655172   
Arizona                  53263  2852.200000   85566     54167  2870.100000   
Arkansas                  5759   257.360000   38604      5606   253.193333   
California              132643  4288.198276  497431    129151  4307.043103   
Colorado                  9584   505.781250   64740      9218   505.773438   
Connecticut               9224  2303.062500   36849      9070  2258.375000   
Delaware                  4673  1859.666667   11158      4590  1815.000000   
District of Columbia         0  4617.000000    9234         0  4723.500000   
Florida                  31247  1597.283582  214036     31065  1589.582090   

                      ... POPESTIMATE2012 POPESTIMATE2013                      \
                      ...       amplitude       meia_soma      soma amplitude   
STNAME                ...                                                       
Alabama               ...          648869    36048.753731   4830533    650207   
Alaska                ...          297960    12714.517241    737442    300901   
Arizona               ...         3936692   221026.633333   6630799   4006420   
Arkansas              ...          383678    19719.713333   2957957    386138   
California            ...         9969310   331156.275862  38414128  10044031   
Colorado              ...          644983    41180.718750   5271132    654451   
Connecticut           ...          817296   224823.000000   3597168    824207   
Delaware              ...          378600   154225.500000    925353    380144   
District of Columbia  ...               0   324770.000000    649540         0   
Florida               ...         2602939   146227.365672  19594467   2633620   

                     POPESTIMATE2014                     POPESTIMATE2015  \
                           meia_soma      soma amplitude       meia_soma   
STNAME                                                                     
Alabama                 36167.246269   4846411    651815    36261.037313   
Alaska                  12707.689655    737046    299723    12731.586207   
Arizona                224292.766667   6728783   4080699   227602.166667   
Arkansas                19778.900000   2966835    387302    19854.693333   
California             334416.301724  38792291  10108329   337455.327586   
Colorado                41840.531250   5355588    663268    42629.484375   
Connecticut            224672.625000   3594762    828986   224430.375000   
Delaware               155994.666667    935968    381574   157655.666667   
District of Columbia   329918.000000    659836         0   336114.000000   
Florida                148549.022388  19905569   2660

Existem outras formas de nomear, mas elas são muito mais complexas e eu acho que não vale a pena aprender.

Entretanto, recomendo você dar uma [olhada](https://pandas.pydata.org/docs/user_guide/groupby.html#named-aggregation).

### Agregações diferentes para colunas diferentes
Para finalizar as agregações, outra coisas que a função `agg` nos permite é aplicar agregações por coluna.

Note que eu passei três funções para `agg` e elas foram aplicadas em todas as colunas numéricas do meu `DataFrame`. Muitas vezes, não queremos isso. Pode ser que eu queira a soma dos nascimentos e a média e amplitude da quantidade de população (diferentes agregações para diferentes colunas). 

Podemos fazer isso passando um dicionário para `agg`:

In [34]:
minhas_agg = {'BIRTHS2013' : np.sum,
              'BIRTHS2014' : np.sum,
              'BIRTHS2015' : np.sum,
              'POPESTIMATE2013' : [np.mean, lambda x : x.max() - x.min()],
              'POPESTIMATE2014' : [np.mean, lambda x : x.max() - x.min()],
              'POPESTIMATE2015' : [np.mean, lambda x : x.max() - x.min()]}
g.agg(minhas_agg).head(10)

BIRTHS2013 BIRTHS2014 BIRTHS2015 POPESTIMATE2013  \
                            sum        sum        sum            mean   
STNAME                                                                  
Alabama                   57938      58334      58305    72097.507463   
Alaska                    11348      11462      11498    25429.034483   
Arizona                   86103      86556      87385   442053.266667   
Arkansas                  37979      37912      37820    39439.426667   
California               499617     499583     503849   662312.551724   
Colorado                  64739      66159      67057    82361.437500   
Connecticut               36134      36182      36225   449646.000000   
Delaware                  10890      10927      11033   308451.000000   
District of Columbia       9447       9415       9593   649540.000000   
Florida                  213004     218088     220628   292454.731343   

                                POPESTIMATE2014            POPESTIMATE2015  \
                     <lambda_0>            mean <lambda_0>            mean   
STNAME                                                                       
Alabama                  650207    72334.492537     651815    72522.074627   
Alaska                   300901    25415.379310     299723    25463.172414   
Arizona                 4006420   448585.533333    4080699   455204.333333   
Arkansas                 386138    39557.800000     387302    39709.386667   
California             10044031   668832.603448   10108329   674910.655172   
Colorado                 654451    83681.062500     663268    85258.968750   
Connecticut              824207   449345.250000     828986   448860.750000   
Delaware                 380144   311989.333333     381574   315311.333333   
District of Columbia          0   659836.000000          0   672228.000000   
Florida                 2633620   297098.044776    2660540   302556.298507   

                                 
                     <lambda_0>  
STNAME                           
Alabama                  651888  
Alaska                   298082  
Arizona                 4158418  
Arkansas                 387435  
California             10169182  
Colorado                 681844  
Connecticut              831480  
Delaware                 383246  
District of Columbia          0  
Florida                 2684786

Veja como obtivemos apenas as somas para os nascimentos e apenas a média e a amplitude para a quantidade de populão.

Além disso, perceba como apenas as colunas que eu informei apareceram (as colunas de 2010, 2011 e 2012 foram ignoradas).

Por fim, note que para ter mais de uma função por coluna, novamente usamos uma lista, como fazíamos antes.

Mas novamente tivemos um problema: a função lambda está sem nome. Será que você consegue resolver isso? 

**Exercício:** 
*Pegue tudo o que já aprendemos e crie uma agregação similar a anterior, mas com nomes 'soma', 'media' e 'amplitude'.*